dbms/src/Storages/StorageMergeTree.cpp

```diff
@@ -1121,4 +1124,54 @@ ActionLock StorageMergeTree::getActionLock(StorageActionBlockType action_type)
     return {};
 }
 
+CheckResults StorageMergeTree::checkData(const ASTPtr & query, const Context & context)
+{
+    CheckResults results;
+    DataPartsVector data_parts;
+    if (const auto & check_query = query->as<ASTCheckQuery &>(); check_query.partition)
+    {
+        String partition_id = getPartitionIDFromQuery(check_query.partition, context);
+        data_parts = getDataPartsVectorInPartition(MergeTreeDataPartState::Committed, partition_id);
+    }
+    else
+        data_parts = getDataPartsVector();
+
+    for (auto & part : data_parts)
+    {
+        String full_part_path = part->getFullPath();
+        /// If the checksums file is not present, calculate the checksums and write them to disk.
+        String checksums_path = full_part_path + "checksums.txt";
+        if (!Poco::File(checksums_path).exists())
+        {
+            try
+            {
+                auto counted_checksums = checkDataPart(part, false, primary_key_data_types, skip_indices);
+                counted_checksums.checkEqual(part->checksums, true);
+                WriteBufferFromFile out(full_part_path + "checksums.txt.tmp", 4096);
```

Comment:
If I understand correctly, this code relies on O_CREAT for synchronization (otherwise race condition will happen).
Need a comment. BTW, who will clean failed `checksums.txt.tmp`?